libtenzir/builtins/operators/to_opensearch.cpp

```diff
@@ -0,0 +1,519 @@
+        .omit_nulls = not args_.include_nulls,
+        .omit_empty_records = false,
+        .omit_empty_lists = false,
+      },
+      args_.max_content_length->inner,
+      args_.compress.has_value(),
+    };
+    for (auto last = time::clock::now(); auto&& slice : input) {
+      const auto now = time::clock::now();
+      if (now - last > args_.buffer_timeout->inner and b.has_contents()) {
+        send_req(req, b.yield(dh), dh);
+        last = now;
+      }
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      // XXX: Why?
+      slice = resolve_enumerations(std::move(slice));
+      constexpr auto ng = []() -> generator<std::optional<std::string_view>> {
+        co_yield {std::nullopt};
+      };
+      const auto ids = resolve_str("id", args_.id, slice, dh);
+      const auto idxs = resolve_str("index", args_.index, slice, dh);
+      const auto acts = resolve_str("action", args_.action, slice, dh);
+      const auto docs = eval(
+        args_.doc.value_or(ast::this_{args_.operator_location}), slice, dh);
+      constexpr auto ty = string_type{};
+      auto id = ids ? values(ty, as<arrow::StringArray>(*ids->array)) : ng();
+      auto idx = idxs ? values(ty, as<arrow::StringArray>(*idxs->array)) : ng();
+      auto act = acts ? values(ty, as<arrow::StringArray>(*acts->array)) : ng();
+      auto doc = docs.values();
+      for (auto&& doc : docs.values()) {
+        const auto ptr = try_as<view<record>>(doc);
+        if (not ptr) {
+          diagnostic::warning("`doc` evaluated to non-record, skipping event")
+            .primary(*args_.doc)
+            .emit(dh);
+          continue;
+        }
+        const auto action = act.next();
+        if (not action or not action.value()) {
+          diagnostic::warning("`action` evaluated to `null`, skipping event")
+            .primary(args_.action)
+            .emit(dh);
+          continue;
+        }
+        if (auto diag
+            = b.create_metadata(**action, idx.next(), id.next(), args_)) {
+          std::move(diag).value().primary(args_.operator_location).emit(dh);
+          continue;
+        }
+        b.create_doc(**action, *ptr);
+        switch (b.finish_event()) {
+          using enum json_builder::state;
+          case ok: {
+            break;
+          }
+          case full: {
+            send_req(req, b.yield(dh), dh);
+            break;
+          }
+          case event_too_large: {
+            diagnostic::warning(
+              "event too large for given `max_content_length`")
+              .note("serialized event size was `{}`", b.last_element_size())
+              .primary(*args_.max_content_length)
+              .emit(dh);
+            break;
+          }
+        }
+      }
+    }
+    if (b.has_contents()) {
+      send_req(req, b.yield(dh), dh);
+    }
+  }
+
+  auto
+  optimize(expression const&, event_order) const -> optimize_result override {
+    return do_not_optimize(*this);
```

Comment:
@dominiklohmann Maybe we want to push an unordered optimization here? I know we also do this in some other cases where its not strictly correct.

(If we do it here, we may also just do it in `to_splunk`)

Reply:
We can only push this if the id field is unset, no? If there is an id field, then the order matters.

Reply:
I am not sure I follow. The ID would be extracted from each event so there wouldn't be an issue.

The only observable effects would be in the randomly generated IDs on the side of opensearch (if you use `action="create" and dont specify an ID, one is generated) and in the ingest time of specific events (which is technically a thing for Splunk I think). But I dont see how the order in which events arrive at opensearch or splunk matters to a user.

Reply:
If the id is not random, then the order of events with matching ids matters for upserting.